prog/vnet/subnet_nexus.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+class Prog::Vnet::SubnetNexus < Prog::Base
+  semaphore :refresh_mesh, :destroy
+
+  def self.assemble(project_id, name: nil, location: "hetzner-hel1", ipv6_range: nil, ipv4_range: nil)
+    project = Project[project_id]
+    unless project || Config.development?
+      fail "No existing project"
+    end
+
+    ubid = UBID.generate(UBID::TYPE_PRIVATE_SUBNET)
+    name ||= PrivateSubnet.ubid_to_name(ubid)
+
+    Validation.validate_name(name)
+
+    ipv6_range ||= random_private_ipv6.to_s
+    ipv4_range ||= random_private_ipv4.to_s
+    DB.transaction do
+      PrivateSubnet.create(name: name, location: location, net6: ipv6_range, net4: ipv4_range, state: "waiting") { _1.id = ubid.to_uuid }
+      Strand.create(prog: "Vnet::SubnetNexus", label: "wait") { _1.id = ubid.to_uuid }
+    end
+  end
+
+  def ps
+    @ps ||= PrivateSubnet[strand.id]
+  end
+
+  def wait
+    when_destroy_set? do
+      hop :destroy
+    end
+
+    when_refresh_mesh_set? do
+      ps.update(state: "refreshing_mesh")
+      hop :refresh_mesh
+    end
+
+    nap 30
+  end
+
+  def refresh_mesh
+    ps.nics.map(&:incr_refresh_mesh)
+    hop :wait_refresh_mesh
+  end
+
+  def wait_refresh_mesh
+    unless ps.nics.any? { SemSnap.new(_1.id).set?("refresh_mesh") }
+      ps.update(state: "waiting")
+      decr_refresh_mesh
+      hop :wait
+    end
+
+    nap 1
+  end
+
+  def destroy
+    if ps.nics.any? { |n| !n.vm_id.nil? }
+      fail "Cannot destroy subnet with active nics, first clean up the attached resources"
+    end
+
+    if ps.nics.empty?
+      ps.delete
```

Comment:
Should we do any cleanup at dataplane?

Reply:
Nope, private subnets are just logically existing entities.